### PR TITLE
Katie and Leanne's Number One Rideshare

### PR DESCRIPTION
**Pull request by katietj**
# Rideshare-Rails
Congratulations! You're submitting your assignment! These comprehension questions should be answered by both partners together, not by a single teammate.

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe the types of entity relationships you set up in your project and why you set up the relationships that way | We created Drivers, Trips, and Passenger models. Passengers  and Drivers are connected to each other through the intermediary model, Trips. 
Describe the role of model validations in your application | We used model validations to ensure the data entered in forms was complete and in the correct format. 
How did your team break up the work to be done? | We did a lot of our work together on the overall programing and code together. Once we got to design and the smaller details of the project, we divided the work up. 
What features did you choose to prioritize in your project, and what features, if any, did you have to set aside to meet the deadline? | We got the basics code aspects done first and then did the secondary functions done after we were sure our site was working without bugs. After we were satisfied with our code, we started on the layout and design.  We didn't get to all of our design details like a hamburger nav or background images.
What was one thing that your team collectively gained more clarity on after completing this assignment? | We gained clarity on how models work together, validations, and forms. 
What is your Trello board URL? | https://trello.com/b/fLwppdZ8/katie-leanne
What is the Heroku URL of your deployed application? | https://number-one-best-rideshare.herokuapp.com/
What are two discussion points that you and your pair discussed when giving/receiving feedback from each other that you would be willing to share? | We had a good pair experience overall! Our communication style worked well for both of us and we were able to bounce ideas off of each other and it made our code better as a result.

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -0,0 +1,50 @@
+Rails.application.routes.draw do
+
```
Comment (CheezItMan) [inline config/routes.rb:2]: Suggestion, you should use resources except for specific custom routes.

File: app/models/driver.rb [hunk 2]
```diff
@@ -0,0 +1,37 @@
+class Driver < ApplicationRecord
+  has_many :trips
+  validates :name, presence: true
+  validates :vin, presence: true, uniqueness: true
```
Comment (CheezItMan) [inline app/models/driver.rb:4]: I suggest also validating that the length is 17.

File: app/models/driver.rb [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    self.trips.each do |trip|
+      total += trip.cost - 1.65
+    end
+    return (total * 0.80).round(2)
+  end
+
+  def ratings
+    average = 0
+    self.trips.each do |trip|
+      average += trip.rating.to_f #to float to allow for half ratings
```
Comment (CheezItMan) [inline app/models/driver.rb:26]: What about incomplete trips?  You need some way to filter out incomplete trips. 

I realize you are creating trips with a rating of 1, but if trips were created without a rating...

File: app/controllers/trips_controller.rb [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+  def new
+    @trip = Trip.new
+  end
+
+  def create
+    available_drivers = Driver.where(status: true)
+    offset = rand(available_drivers.count)
+    driver = available_drivers.offset(offset).first
+    new_trip = {driver_id: driver.id, passenger_id: params[:passenger_id].to_i, date: DateTime.now, cost: rand(500..3000),rating: 1}
```
Comment (CheezItMan) [inline app/controllers/trips_controller.rb:37]: So new trips always start with a rating of 1?  That would make for a lot of unhappy drivers.

File: app/models/trip.rb [hunk 5]
```diff
@@ -0,0 +1,15 @@
+class Trip < ApplicationRecord
+  belongs_to :driver
+  belongs_to :passenger
+  validates :rating, presence: true, format: { with: /^[1-5]$/, multiline: true} #multiline to allow the start karat and end dollar sign in regex
+
+  #task is independent of context -- biz logic only
+  def earnings
```
Comment (CheezItMan) [inline app/models/trip.rb:7]: I'm not sure what this method is for.  You're calculating the total earnings for all the trips, regardless of driver.

**Comment by CheezItMan** [comment]
# Rideshare Rails
## What We're Looking For

Feature | Feedback
:------------- | :-------------
**Baseline** |
Appropriate git usage with no extraneous files checked in and both partners contributing |  Good number of commits and good commit messages, both partners contributed.  
Answered comprehension questions |  Check, glad you two worked together well. 
Uses named routes (like `_path`) |  Check
RESTful routes utilized |  Yes well done, but I suggest using resources unless you need specific custom routes.  
**Project Requirements** |
Table relationships |  Check
Validation rules for Models |  Check
Business logic is in the models |  Check
Database is seeded from the CSV files |  Check
Trello board is created and utilized in project management |Check
Heroku instance is online |  Check
The app is styled to create an attractive user interface |  That orange is sooo loud it hurts my brain.  
**Overall** |  You hit all the requirement for the project.  Nice work and well done!

File: app/controllers/trips_controller.rb (hunk 4 above)
Comment (leannerivera) [reply app/controllers/trips_controller.rb:37]: When the trip ends, the user must rate the trip.

File: app/models/trip.rb (hunk 5 above)
Comment (leannerivera) [reply app/models/trip.rb:7]: It works on our Driver list page where they are shown their total earnings.

**Pull request closed without merging**